### PR TITLE
Optimize normalization

### PR DESCRIPTION
**Pull request by martinbonnin**
Introduce `NormalizationIRResponseWriter`, a ResponseWriter that writes a "normalization IR" later used to compute Records. 

This removes one intermediate `Map` and tree traversal an makes the `apolloParseAndNormalize` benchmark from ~30 to ~22 ms.

A nice side effect is that it fixes computing records for Json scalar which didn't work when normalizing models (it works in main when normalizing the network response, see https://github.com/apollographql/apollo-android/pull/2859)

### CODE REVIEW

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.apollographql.apollo.cache.normalized.internal
+
+import com.apollographql.apollo.api.ResponseField
+
+/**
+ * A hierarchy of Json object with their fields attached used for normalization
+ */
+sealed class AJElement
```
Comment (sav007) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt:8]: I would recommend to change the name to smth different, as `AJ` is so abstract and not very descriptive for `augmented json`. Couple suggestion as this only related to cache normalization maybe `NormalizationNode/ Element`, `CacheNormalizationNode / Element`?

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJResponseWriter.kt [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+  private var rootObject: AJObject? = null
+  fun root(): AJObject {
+    if (rootObject == null) {
+      /**
+       * close the root object  if needed
+       */
+      check(objectStack.size == 1)
+      rootObject = AJObject(fields = objectStack.removeLast())
+    }
```
Comment (sav007) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJResponseWriter.kt:31]: not sure but is it possible to use `lazy` intializer here?

Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJResponseWriter.kt]: Yup, way more idiomatic

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+class AJScalar(val value: Any?) : AJElement()
+
+fun AJElement.mergeWith(other: AJElement): AJElement {
+  return when {
+    this is AJScalar && other is AJScalar -> {
+      check(this.value == other.value)
+      this
+    }
+    this is AJList && other is AJList -> throw UnsupportedOperationException()
```
Comment (sav007) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt:23]: hm why is that? We can't merge lists?

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt]: Renamed to `NormalizationIR`

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt (hunk 3 above)
Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJ.kt]: Not sure why I did that, I guess we need to merge lists nested in arrays indeed. I just added it.

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/NormalizationIR.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package com.apollographql.apollo.cache.normalized.internal
+
+import com.apollographql.apollo.api.ResponseField
+
+
+/**
+ * A hierarchy of Json object with their fields attached used for normalization
+ */
+internal object NormalizationIR {
+  sealed class Element {
```
Comment (sav007) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/NormalizationIR.kt:10]: up to you, just a thought if `Node` suit better?

**Review by sav007 — APPROVED** [review]

**Pull request merged**